lib/mayaUsd/ufe/UsdStageMap.cpp

```diff
@@ -85,32 +110,53 @@ void UsdStageMap::addItem(const Ufe::Path& path, UsdStageWeakPtr stage)
         return;
     }
 
-    // Convert the tail of the UFE path to an MObjectHandle.
-    auto proxyShape = proxyShapeHandle(path);
+    // Convert the UFE path to an MObjectHandle.
+    auto proxyShape = nameLookup(path);
     if (!proxyShape.isValid()) {
         return;
     }
 
-    // Could get the stage from the proxy shape object in the stage() method,
-    // but since it's given here, simply store it.
-    fObjectToStage[proxyShape] = stage;
+    // We've just done a name-based lookup of the proxy shape, so the stage
+    // cannot be null.
+    auto stage = objToStage(proxyShape.object());
+    TF_AXIOM(stage);
+
+    fPathToObject[path] = proxyShape;
     fStageToObject[stage] = proxyShape;
 }
 
-UsdStageWeakPtr UsdStageMap::stage(const Ufe::Path& path)
+UsdStageWeakPtr UsdStageMap::stage(const Ufe::Path& path) { return objToStage(proxyShape(path)); }
+
+MObject UsdStageMap::proxyShape(const Ufe::Path& path)
 {
+    const auto& singleSegmentPath
+        = nbPathSegments(path) == 1 ? path : Ufe::Path(path.getSegments()[0]);
+
+    auto iter = fPathToObject.find(singleSegmentPath);
+    if (iter != std::end(fPathToObject)) {
+        return iter->second.object();
+    }
+
+    // Did not find object in cache, either because the cache is stale, or
+    // because the object does not exist.  Rebuild the cache and try again.
+    setDirty();
     rebuildIfDirty();
 
-    auto proxyShape = proxyShapeHandle(path);
-    if (!proxyShape.isValid()) {
+    // At this point the cache is rebuilt, so lookup failure means the object
+    // doesn't exist.
+    iter = fPathToObject.find(singleSegmentPath);
+    return iter == std::end(fPathToObject) ? MObject() : iter->second.object();
+}
+
+MayaUsdProxyShapeBase* UsdStageMap::proxyShapeNode(const Ufe::Path& path)
```

Comment:
New accessor to the cache for the proxy shape node pointer.  Allows us to retrieve data from the proxy shape node, such as time, or enabled purposes.